extensions/amp-app-banner/0.1/amp-app-banner.js

```diff
@@ -0,0 +1,177 @@
+    } else if (platform.isAndroid()) {
+      this.preconnect.url('https://play.google.com', onLayout);
+    }
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private @const {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
+    if (!this.isExperimentOn_) {
+      user.warn(TAG, `Experiment ${TAG} disabled`);
+      return;
+    }
+
+    if (!platform.isIos() && !platform.isAndroid()) {
+      setStyles(this.element, {
+        'display': 'none',
+      });
+      return;
+    }
+
+    /** @private @const {!Viewer} */
+    const viewer = viewerFor(this.getWin());
+
+    /** @private @const {!Document} */
+    this.doc_ = this.getWin().document;
+
+    /** @private @const {!Xhr} */
+    this.xhr_ = xhrFor(this.getWin());
+
+    /** @const @private {!Element} */
+    this.installLink_ = this.element.querySelector('a[install-link]');
+    user.assert(this.installLink_, '<a install-link> is required inside %s: %s',
+        TAG, this.element);
+
+    /** @const @private {!Element} */
+    this.openLink_ = this.element.querySelector('a[open-link]');
+    user.assert(this.openLink_, '<a open-link> is required inside %s: %s',
+        TAG, this.element);
+
+    if (platform.isIos()) {
+      const meta = this.doc_.head.querySelector(
+          'meta[name=apple-itunes-app]');
+      user.assert(meta,
+          '<meta name=apple-itunes-app> in <head> is required: %s',
+          this.element);
+
+      // If non-embedded document and on Safari iOS just use the native app
+      // banner Safari supports if the meta tag is already setup.
+      if (!viewer.isEmbedded() && platform.isSafari()) {
+        setStyles(this.element, {
+          'display': 'none',
+        });
+        return;
+      }
+      this.parseIosMetaContent_(meta.getAttribute('content'));
+    }
+
+    // TODO: Provide a way to dismiss the banner and persist it.
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (!this.isExperimentOn_) {
+      user.warn(TAG, `Experiment ${TAG} disabled`);
+      return Promise.resolve();
+    }
+
+    if (!platform.isAndroid()) {
+      return Promise.resolve();
+    }
+
+    const manifestLink = this.doc_.head.querySelector(
+        'link[rel=manifest],link[rel=amp-manifest]');
+    user.assert(manifestLink,
+        '<link rel=manifest> in the <head> is required: %s',
+        this.element);
+
+    const manifestHref = manifestLink.getAttribute('href');
+    assertHttpsUrl(manifestHref, this.element);
+
+    return this.xhr_.fetchJson(manifestHref)
+        .then(response => {
+          this.parseManifest_(response);
+        }).catch(unusedError => {
+          // TODO: What do we do when we fail to fetch manifest.
+          // Hiding it at this point might jump the page?
+        });
+  }
+
+  /**
+   * @param {!Object} manifestJson
+   * @private
+   */
+  parseManifest_(manifestJson) {
```

Comment:
Manifests aren't supported due to: https://github.com/ampproject/amphtml/issues/2975